x-pack/plugins/ml/server/models/data_recognizer/data_recognizer.ts

```diff
@@ -247,6 +252,75 @@ export class DataRecognizer {
     return configs.find((i) => i.module.id === id);
   }
 
+  // called externally by an endpoint
+  public async findIndexMatches(
+    moduleId: string,
+    size: number = 15
+  ): Promise<RecognizeModuleResultDataView[]> {
+    const config = await this._findConfig(moduleId);
+    const matchingDataViews: RecognizeModuleResultDataView[] = [];
+
+    if (config?.module.query === undefined) {
+      return matchingDataViews;
+    }
+
+    try {
+      const promises: Array<
+        Promise<SearchResponse<unknown, Record<string, AggregationsAggregate>>>
+      > = [];
+      // Keep track of which data view title is associated with which promise
+      const dataViewTitleMap: Record<string, Promise<any>> = {};
+      const dataViewsMap: Record<string, { id: string; name: string | undefined }> = {};
+      const idsWithTitle = await this._dataViewsService.getIdsWithTitle();
+
+      idsWithTitle.forEach(({ id, title, name }) => {
+        dataViewsMap[title] = { id, name };
+        const promise = this._client.asCurrentUser.search(
+          {
+            index: title, // title is index pattern we can search by e.g. "filebeat-*"
+            size: 0,
+            body: {
+              query: config?.module.query,
+            },
+          },
+          { maxRetries: 0 }
+        );
+        dataViewTitleMap[title] = promise;
+        promises.push(promise);
+      });
+
+      const response = await Promise.all(promises);
+
+      const responseMap: Record<string, number> = Object.keys(dataViewTitleMap).reduce(
+        (acc, title, i) => {
+          const totalHits =
+            typeof response[i].hits.total === 'number'
+              ? response[i].hits.total
+              : (response[i].hits?.total as SearchTotalHits)?.value ?? 0;
+          return Object.assign(acc, { [title]: totalHits });
+        },
+        {}
+      );
+
+      for (const title in responseMap) {
+        if (Object.prototype.hasOwnProperty.call(responseMap, title)) {
+          const indexResponseHitsTotal = responseMap[title];
+          if (indexResponseHitsTotal > 0 && matchingDataViews.length < size) {
+            matchingDataViews.push({ title, ...(dataViewsMap[title] ?? {}) });
+          } else if (matchingDataViews.length >= size) {
+            break;
+          }
+        }
+      }
```

Comment:
I had a quick go at writing this logic without using any maps

```suggestion
      const idsWithTitle = await this._dataViewsService.getIdsWithTitle();

      // create temp objects with a function for running the query
      const tempObjs = idsWithTitle.map(({ id, title, name }) => {
        return {
          id,
          title,
          name,
          func: async () =>
            this._client.asCurrentUser.search(
              {
                index: title, // title is index pattern we can search by e.g. "filebeat-*"
                size: 0,
                body: {
                  query: config?.module.query,
                },
              },
              { maxRetries: 0 }
            ),
        };
      });
      // run all the queries in parallel
      const response = await Promise.all(
        tempObjs.map(async ({ id, name, title, func }) => {
          const resp = await func();
          const totalHits =
            typeof resp.hits.total === 'number'
              ? resp.hits.total
              : (resp.hits?.total as SearchTotalHits)?.value ?? 0;
          return { id, name, title, totalHits };
        })
      );

      // filter and return only those with hits
      return response.filter((r) => r.totalHits > 0).slice(0, size);
```

Reply:
🤔 I like the efficiency but find it a bit less readable. Good to remove the creation of the temporary maps but still the same number of loops through and unable to break out of the last loop when size has been met. Both have pros/cons - I went ahead and updated, though, as the optimization of the removal of the maps might be worth it. 👍